Managers/PerformanceMan.h

```diff
@@ -0,0 +1,215 @@
+
+		/// <summary>
+		/// Saves current absolute time in microseconds as an end of performance measurement. The difference is added to the value of current performance sample.
+		/// </summary>
+		/// <param name="counter">Counter to stop and updated measurement for.</param>
+		void StopPerformanceMeasurement(PerformanceCounters counter);
+
+		/// <summary>
+		/// Sets the current ping value to display.
+		/// </summary>
+		/// <param name="ping">Ping value to display.</param>
+		void SetCurrentPing(unsigned int ping) { m_CurrentPing = ping; }
+#pragma endregion
+
+#pragma region Class Info
+		/// <summary>
+		/// Gets the class name of this Entity.
+		/// </summary>
+		/// <returns>A string with the friendly-formatted type name of this object.</returns>
+		virtual const std::string & GetClassName() const { return c_ClassName; }
+#pragma endregion
+
+	protected:
+
+		static const std::string c_ClassName; //!< A string with the friendly-formatted type name of this object.
+		static constexpr unsigned short c_MSPFAverageSampleSize = 10;
+		static constexpr unsigned short c_MaxSamples = 120; //!< How many performance samples to store, directly affects graph size.
+		static constexpr unsigned short c_Average = 10; //!< How many samples to use to calculate average value displayed on screen.
+
+		const unsigned short c_StatsOffsetX = 17; //!< Offset of the stat text from the left edge of the screen.
+		const unsigned short c_StatsHeight = 14; //!< Height of each stat text line.
+		const unsigned short c_GraphsOffsetX = 14; //!< Offset of the graph from the left edge of the screen.
+		const unsigned short c_GraphsStartOffsetY = 134; //!< Position the first graph block will be drawn from the top edge of the screen.
+		const unsigned short c_GraphHeight = 20; //!< Height of the performance graph.
+		const unsigned short c_GraphBlockHeight = 34; //!< Height of the whole graph block (text height and graph height combined).
+
+		bool m_ShowPerfStats; //!< Whether to show performance stats on screen or not.
+		bool m_AdvancedPerfStats; //!< Whether to show performance graphs on screen or not.
+
+		Timer *m_FrameTimer; //!< Timer for measuring millisecs per frame for performance stats readings.
+		float m_SimSpeed; //!< The simulation speed over real time.
+		unsigned short m_Sample; //!< Sample counter.
+
+		std::deque<unsigned int> m_MSPFs; //!< History log of readings, for averaging the results.
+		unsigned short m_MSPFAverage; //!< The average of the MSPF reading buffer above, computer each frame.
+		unsigned short m_CurrentPing; //!< Current ping value to display on screen.
+
+		std::string m_PerfCounterNames[PERF_COUNT]; //!< Performance counter's names displayed on screen.
+		unsigned short m_PerfPercentages[PERF_COUNT][c_MaxSamples]; //!< Array to store percentages from PERF_SIM_TOTAL
+		uint64_t m_PerfData[PERF_COUNT][c_MaxSamples]; //!< Array to store performance measurements in microseconds.	
```

Comment:
This is a fixed width integer type, it guarantees to be an unsigned integer of at least 64 bits.
I guess I can replace it with `unsigned long long` but goddamn that is LONG LONG.

Reply:
Yeah, like I said, I actually prefer `uint64_t someVariable` to `unsigned long long long long long long someVariable` because dear god the latter is LONG LONG. I'd like to be consistent though, so it's used anywhere we have `unsigned long long`, that way there's no confusion (maybe define it in constants so we're not dependent on it being defined in whatever library it's coming from?)

Reply:
I think i'm gonna go with `unsigned long long long long` here after all.
Besides this typedef being used here, it is being used in 2 lines (one in `AudioMan` and one in `TimerMan`).
Other than that, `unsigned long long` is only being used once in `AudioMan`.
So all in all this longass thing is only "widely" used in `PerformanceMan` and we can live with it.

The reasoning behind this is that we actually have a conflict of `uint64_t`s and other such typedefs.
One is being "universal" and is defined as typedef in `stdint.h` and the other is a macro defined by Allegro.
The one defined by Allegro is actually using a Windows specific type `__int64`.
In `Constants` we're actually *un*defining all the macros defined in Allegro so defining it there is silly.
Generally in the case of `AudioMan` and `TimerMan`, the Allegro macro will be used because `Constants` isn't included in either, but this isn't an issue.
The issue is that the Allegro macros mess up something in RakNet, causing redefinition errors with WinSock or whatever, I don't remember.

Bottom line is that I don't want to include `Constants` in those 2 or undef them separately so if I use the typedef we end up with supposedly the same but different types of `unsigned long long` between these 3 classes we use it in.

So i'm just gonna avoid all that and `unsigned long long` all the 10 lines we use it in.
I hope this isn't too much of a mess of an explanation.

Reply:
Yeah this is the problem with this sort of thing, and it's why you can't do it in more modern languages. You end up with weirdass shit like this that makes it not worth doing in the first place lol.
Anyway, yeah, I'm fine with that. I agree with your logic.